src/Swift.Runtime/src/Swift/SwiftEquatable.cs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices.Swift;
+using Swift.Runtime.InteropServices;
+
+namespace Swift.Runtime
+{
+    public interface ISwiftEquatable
+    {
+        // This interface is intentionally left empty. It serves as a marker for types that implement Swift's Equatable protocol.
+    }
+
+    /// <summary>
+    /// Provides functionality to use Swift's Equatable protocol for equality comparison.
+    /// </summary>
+    public static class SwiftEquatable
+    {
+        // P/Invoke declaration for Swift's Equatable protocol equality operator
+        [UnmanagedCallConv(CallConvs = new Type[] { typeof(CallConvSwift) })]
+        [DllImport(KnownLibraries.SwiftCore, EntryPoint = "$sSQ2eeoiySbx_xtFZTj")]
+        private static extern bool PInvoke_SwiftEquals(
+            IntPtr lhs,
+            IntPtr rhs,
+            SwiftSelf typeMetadataInSwiftSelf,
+            TypeMetadata typeMetadata,
+            ProtocolWitnessTable equatableProtocolWitnessTable);
+
+        /// <summary>
+        /// Compares two objects using Swift's Equatable protocol.
+        /// </summary>
+        /// <typeparam name="T">Type that implements IEquatable&lt;T&gt;</typeparam>
+        /// <param name="lhs">Left-hand side object</param>
+        /// <param name="rhs">Right-hand side object</param>
+        /// <returns>True if the objects are equal according to Swift's equality</returns>
+        public static unsafe bool Equals<T>(T lhs, T rhs)
+        {
+            if (lhs == null)
+                throw new ArgumentNullException(nameof(lhs));
+            if (rhs == null)
+                throw new ArgumentNullException(nameof(rhs));
+
+            var metadata = TypeMetadata.GetTypeMetadataOrThrow<T>();
+            var equatablePwt = ProtocolWitnessTable.GetOrThrow<T, ISwiftEquatable>();
+
+            IntPtr lhsPayload = IntPtr.Zero;
+            IntPtr rhsPayload = IntPtr.Zero;
+
+            try
+            {
+                lhsPayload = (IntPtr)NativeMemory.Alloc(metadata.Size);
+                rhsPayload = (IntPtr)NativeMemory.Alloc(metadata.Size);
```

Comment:
Is it possible in Swift to compare different types for equality?

Reply:
I dont think you can achieve that without some conversion to common type first

Reply:
Do you know if Swift compiler generates these conversions automatically or is it a user responsibility?

Actually, I just tried simple example with:

```swift
let x: Double = 4.3;
let y: Float = 2.3;

if (x == y) {
...
}
```
and it generates a compiler warning that it requires conversion to common type.